File: examples/julia/station_series.jl
# One station's QC-passed CO2 series in Julia, via the REST route.
#
# Verified against the live service 2026-08-18 with Julia 1.10.4:
#   8701 rows, mean 426.837 ppm — the same numbers R and Python return.
#
# Why REST rather than Zarr.jl: the QC flags are |S1 byte arrays whose
# fill_value is base64 per the Zarr v2 spec, and Zarr.jl tries to convert that
# string to a character. `zopen` reads a group's arrays eagerly, so one flag
# array fails the whole group — you can read the numeric arrays one at a time,
# but never the flags, so client-side QC is not possible. apply_qc=true lets
# the server do it.

using HTTP, JSON3, DataFrames, Dates

q = Dict("id" => "icos-obspack.co2.co2", "station" => "CBW207",
         "start" => "2023-01-01", "end" => "2024-01-01",
         "apply_qc" => "true", "max_rows" => "200000")

body  = String(HTTP.get("https://zarr.icos-cp.eu/query"; query=q).body)
lines = split(body, '\n'; keepempty=false)
rows  = [JSON3.read(l) for l in lines[1:end-1]]   # the last line is the passport

# Type the columns explicitly. Building the DataFrame straight from JSON3
# values gives an abstract `Real` column and timestamps left as `String`.
df = DataFrame(
    station = String[r.station for r in rows],
    time    = DateTime[DateTime(r.time_co2, dateformat"yyyy-mm-ddTHH:MM:SSZ") for r in rows],
    co2     = Float64[r.co2 for r in rows],
)

println("rows: ", nrow(df), "  types: ", eltype.(eachcol(df)))
println("time range (UTC): ", minimum(df.time), " -> ", maximum(df.time))  # end is EXCLUSIVE
println("mean co2: ", round(sum(df.co2) / nrow(df), digits=3), " ppm")

# Cite what you used — the passport is the last line of the stream.
pp   = JSON3.read(lines[end])._passport
root = first(filter(n -> get(n, "@id", "") == "./", pp["@graph"]))
println("citation: ", first(root["citation"], 60), "...")
println("license : ", root["license"]["@id"])

File: examples/octave/station_series.m
% One station's QC-passed CO2 series in GNU Octave.
%
% Verified against the live service 2026-08-18 with Octave 8.4.0 (no packages):
%   8701 rows, mean 426.837 ppm — the same numbers R, Julia and Python return.
%
% Octave is NOT MATLAB for this purpose: `table`, `readtable`, `datetime`,
% `parquetread` and `zarrread` do not exist, and neither does `websave`. What
% does exist is `urlread`/`webread` plus a builtin `jsondecode`, which is
% enough for both routes below.

base = 'https://zarr.icos-cp.eu/query';
q = ['?id=icos-obspack.co2.co2&station=CBW207&start=2023-01-01' ...
     '&end=2024-01-01&apply_qc=true'];

% ---- bulk route: CSV ----------------------------------------------------
% Cheapest to parse, but Octave cannot see response headers, so the citation
% that a CSV response carries in X-Data-Citation is out of reach here.
txt   = urlread([base q '&format=csv']);
lines = strsplit(strtrim(txt), "\n");
C     = textscan(strjoin(lines(2:end), "\n"), '%q %q %f %f %f %q %f', ...
                 'Delimiter', ',');

station = C{1};  co2 = C{3};  flag = C{6};

% No datetime type: use datenum. Strip the fractional seconds first.
dn = datenum(cellfun(@(s) s(1:19), C{2}, 'UniformOutput', false), ...
             'yyyy-mm-dd HH:MM:SS');

printf('rows: %d\n', numel(co2));
printf('time range (UTC): %s -> %s\n', datestr(dn(1)), datestr(dn(end)));
printf('mean co2: %.3f ppm\n', mean(co2));
printf('qc flags kept: %s\n', strjoin(unique(flag)', ','));

% ---- provenance route: ndjson ------------------------------------------
% PREFER THIS for anything you intend to publish. The passport is the last
% line of the stream, so it needs no headers and no Arrow/Parquet support —
% which makes ndjson the only route in Octave that carries provenance.
txt   = urlread([base q]);
lines = strsplit(strtrim(txt), "\n");
rows  = cellfun(@jsondecode, lines(1:end-1), 'UniformOutput', false);
pp    = jsondecode(lines{end})._passport;

printf('\nndjson rows: %d\n', numel(rows));
printf('first: %s  %.3f ppm\n', rows{1}.time_co2, rows{1}.co2);

graph = pp.x_graph;                       % jsondecode maps @graph -> x_graph
for k = 1:numel(graph)
  node = graph{k};
  if isfield(node, 'x_id') && strcmp(node.x_id, './')
    printf('citation: %.60s...\n', node.citation);
    break
  end
end
